File: chartview/src/main/java/com/bitvale/chartview/widget/chart/ChartContainer.java
package com.bitvale.chartview.widget.chart;

import android.content.Context;
import android.graphics.Color;
import android.util.AttributeSet;
import android.util.Log;
import android.view.LayoutInflater;
import android.widget.LinearLayout;
import androidx.annotation.Nullable;
import com.bitvale.chartview.ChartViewListener;
import com.bitvale.chartview.widget.ChipView;
import com.bitvale.chartview.R;
import com.bitvale.chartview.model.Chart;

import java.util.ArrayList;

/**
 * Created by Alexander Kolpakov (jquickapp@gmail.com) on 16-Mar-19
 */
public class ChartContainer extends LinearLayout {

    private ChartView chartView;
    private ChartSpinner chartSpinner;
    private LinearLayout controlsContainer;

    public ChartContainer(Context context) {
        super(context);
        init(context, null, 0);
    }

    public ChartContainer(Context context, @Nullable AttributeSet attrs) {
        super(context, attrs);
        init(context, attrs, 0);
    }

    public ChartContainer(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context, attrs, defStyleAttr);
    }

    private void init(Context context, AttributeSet attrs, int defStyleAttr) {
        setOrientation(LinearLayout.VERTICAL);

        chartView = new ChartView(context, attrs, defStyleAttr);

        chartSpinner = new ChartSpinner(context, attrs, defStyleAttr);
        chartSpinner.setChartListener(chartView);

        controlsContainer = new LinearLayout(context);
        controlsContainer.setOrientation(LinearLayout.HORIZONTAL);

        int margin = context.getResources().getDimensionPixelSize(R.dimen.def_margin);

        addView(chartView);
        addView(chartSpinner);
        ((MarginLayoutParams) chartSpinner.getLayoutParams()).setMargins(0, margin / 2, 0, 0);
        addView(controlsContainer);
        ((MarginLayoutParams) controlsContainer.getLayoutParams()).setMargins(0, margin, 0, 0);
    }

    public void setupData(Chart chart) {
        ArrayList<Long> xAxis = chart.columns.get(0).values;
        ArrayList<Chart.Column> yAxis = new ArrayList<>();
        for (int i = 1; i < chart.columns.size(); i++) {
            yAxis.add(chart.columns.get(i));
        }
        chartView.setupData(xAxis, yAxis);
        chartSpinner.setupData(xAxis, yAxis);
        createControls(chart);
    }

    private void createControls(Chart chart) {
        controlsContainer.removeAllViews();
        LayoutInflater inflater = LayoutInflater.from(getContext());
        for (int i = 1; i < chart.columns.size(); i++) {
            ChipView chip = (ChipView) inflater.inflate(R.layout.chip_view, this, false);
            Chart.Column column = chart.columns.get(i);
            chip.setChecked(column.enabled);
            chip.setText(column.name);
            chip.setCheckedColor(Color.parseColor(column.color));
            chip.setOnClickListener(v -> {
                int enabled = 0;
                if (chip.isChecked()) {
                    for (int j = 1; j < chart.columns.size(); j++) {
                        if (chart.columns.get(j).enabled) {
                            enabled++;
                        }
                    }
                }
                if (enabled == 1) return;

                column.enabled = !column.enabled;
                Chart.ChartAnimation animation = Chart.ChartAnimation.DOWN;
                if (chip.isChecked()) animation = Chart.ChartAnimation.UP;
                column.animation = animation;

                chartView.animateInOut(chip.isChecked());
                chartSpinner.animateInOut(chip.isChecked());
                chip.animateChecked();
                chartSpinner.invalidate();
            });
            controlsContainer.addView(chip);
        }
    }

    public void setChartViewListener(ChartViewListener listener) {
        chartView.setChartViewListener(listener);
    }
}

File: chartview/src/main/java/com/bitvale/chartview/widget/chart/ChartContainer.kt
package com.bitvale.chartview.widget.chart

import android.content.Context
import android.graphics.Color
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.LinearLayout
import androidx.core.view.updateMargins
import com.bitvale.chartview.model.Chart
import com.bitvale.chartview.ChartViewListener
import com.bitvale.chartview.widget.ChipView
import com.bitvale.chartview.R

/**
 * Created by Alexander Kolpakov (jquickapp@gmail.com) on 10-Mar-19
 */
class ChartContainer @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : LinearLayout(context, attrs, defStyleAttr) {

    private var chartView = ChartView(context, attrs, defStyleAttr)
    private var chartSpinner = ChartSpinner(context, attrs, defStyleAttr)
    private var controlsContainer = LinearLayout(context).apply {
        orientation = LinearLayout.HORIZONTAL
    }
    private val margin = context.resources.getDimensionPixelSize(R.dimen.def_margin)

    init {
        orientation = LinearLayout.VERTICAL
        chartSpinner.setChartListener(chartView)
        addView(chartView)
        addView(chartSpinner)
        (chartSpinner.layoutParams as MarginLayoutParams).updateMargins(top = margin / 2)
        addView(controlsContainer)
        (controlsContainer.layoutParams as MarginLayoutParams).updateMargins(top = margin)
    }

    fun setupData(chart: Chart) {
        val xAxis = chart.columns[0].values
        val yAxis = ArrayList<Chart.Column>()
        for (i in 1 until chart.columns.size) {
            yAxis.add(chart.columns[i])
        }
        chartView.setupData(xAxis, yAxis)
        chartSpinner.setupData(xAxis, yAxis)
        createControls(chart)
    }

    private fun createControls(chart: Chart) {
        controlsContainer.removeAllViews()
        val inflater = LayoutInflater.from(context)
        for (i in 1 until chart.columns.size) {
            val chip: ChipView = inflater.inflate(R.layout.chip_view, this, false) as ChipView
            val column = chart.columns[i]
            chip.isChecked = column.enabled
            chip.setText(column.name)
            chip.setCheckedColor(Color.parseColor(column.color))
            chip.setOnClickListener {
                var enabled = 0
                if (chip.isChecked) {
                    for (j in 1 until chart.columns.size) {
                        if (chart.columns[j].enabled) {
                            enabled++
                        }
                    }
                }
                if (enabled == 1) return@setOnClickListener

                column.enabled = !column.enabled
                column.animation = if (chip.isChecked) Chart.ChartAnimation.UP else Chart.ChartAnimation.DOWN
                chartView.animateInOut(chip.isChecked)
                chip.animateChecked()
                chartSpinner.invalidate()
            }
            controlsContainer.addView(chip)
        }
    }

    fun setChartViewListener(listener: ChartViewListener) {
        chartView.setChartViewListener(listener)
    }
}